pyomo/contrib/piecewise/piecewise_linear_function.py

```diff
@@ -0,0 +1,357 @@
+from pyomo.core.base.initializer import Initializer
+import pyomo.core.expr.current as EXPR
+
+np, numpy_available = attempt_import('numpy')
+scipy, scipy_available = attempt_import('scipy')
+spatial, scipy_available = attempt_import('scipy.spatial')
+
+# This is the default absolute tolerance in np.isclose... Not sure if it's
+# enough, but we need to make sure that 'barely negative' values are assumed to
+# be zero.
+ZERO_TOLERANCE = 1e-8
+
+class PiecewiseLinearFunctionData(_BlockData):
+    _Block_reserved_words = Any
+
+    def __init__(self, component=None):
+        _BlockData.__init__(self, component)
+
+        with self._declare_reserved_components():
+            self._expressions = Expression(NonNegativeIntegers)
+            self._simplices = None
+            # These will always be tuples, even when we only have one dimension.
+            self._points = []
+            self._linear_functions = []
+
+    def __call__(self, *args):
+        if all(type(arg) in EXPR.native_types or not
+               arg.is_potentially_variable() for arg in args):
+            # We need to actually evaluate
+            return self._evaluate(*args)
+        else:
+            idx = len(self._expressions)
+            expr = PiecewiseLinearExpression(args, self, idx)
+            self._expressions[idx] = expr
+            return self._expressions[idx]
+
+    def _evaluate(self, *args):
+        # ESJ: This is a very inefficient implementation in high dimensions, but
+        # for now we will just do a linear scan of the simplices.
+        if self._simplices is None:
+            raise RuntimeError("Cannot evaluate PiecewiseLinearFunction--it "
+                               "appears it is not fully defined. (No simplices "
+                               "are stored.)")
+
+        pt = [value(arg) for arg in args]
+        for simplex, func in zip(self._simplices, self._linear_functions):
+            if self._pt_in_simplex(pt, simplex):
+                return func(*args)
+
+        raise ValueError("Unsuccessful evaluation of PiecewiseLinearFunction "
+                         "'%s' at point (%s). Is the point in the function's "
+                         "domain?" %
+                         (self.name, ', '.join(str(arg) for arg in args)))
+
+    def _pt_in_simplex(self, pt, simplex):
+        dim = len(pt)
+        if dim == 1:
+            return self._points[simplex[0]][0] <= pt[0] and \
+                self._points[simplex[1]][0] >= pt[0]
+        # Otherwise, we check if pt is a convex combination of the simplex's
+        # extreme points
+        A = np.ones((dim + 1, dim + 1))
+        b = np.array([x for x in pt] + [1])
+        for j, extreme_point in enumerate(simplex):
+            for i, coord in enumerate(self._points[extreme_point]):
+                A[i, j] = coord
+        try:
+            lambdas = np.linalg.solve(A, b)
+        except np.linalg.LinAlgError as e:
+            if 'Singular' in str(e):
```

Comment:
Ohhh, they probably don't have a separate error because we can just check the determinant of A. I'm going to do that. Then if solving the system throws an error, there really is a problem--I don't have to catch anything.